source/opt/loop_unroller.cpp

```diff
@@ -420,36 +491,58 @@ void LoopUnrollerUtilsImpl::PartiallyUnrollResidualFactor(ir::Loop* loop,
       *context_->GetLoopDescriptor(&function_);
 
   loop_descriptor.AddLoop(new_loop, loop->GetParent());
+
+  RemoveDeadInstructions();
 }
 
-// Duplicate the |loop| body |factor| number of times while keeping the loop
-// backedge intact.
-void LoopUnrollerUtilsImpl::PartiallyUnroll(ir::Loop* loop, size_t factor) {
-  Unroll(loop, factor);
-  AddBlocksToLoop(loop);
-  AddBlocksToFunction(loop->GetMergeBlock());
+// Mark this loop as DontUnroll as it will already be unrolled and it may not
+// be safe to unroll a previously partially unrolled loop.
+void LoopUnrollerUtilsImpl::MarkLoopControlAsDontUnroll(ir::Loop* loop) const {
+  ir::Instruction* loop_merge_inst = loop->GetHeaderBlock()->GetLoopMergeInst();
+  assert(loop_merge_inst &&
+         "Loop merge instruction could not be found after entering unroller "
+         "(should have exited before this)");
+
+  uint32_t loop_control_dont_unroll = 2;
+  uint32_t loop_control_index = 2;
+  loop_merge_inst->SetInOperand(loop_control_index, {loop_control_dont_unroll});
 }
 
-// Duplicate the |loop| body |factor| number of times while keeping the loop
-// backedge intact.
+// Duplicate the |loop| body |factor| - 1 number of times while keeping the loop
+// backedge intact. This will leave the loop with |factor| number of bodies
+// after accounting for the initial body.
 void LoopUnrollerUtilsImpl::Unroll(ir::Loop* loop, size_t factor) {
+  // If we unroll a loop partially it may not be safe to unroll it further.
+  MarkLoopControlAsDontUnroll(loop);
+
+  std::vector<ir::Instruction*> inductions;
+  loop->GetInductionVariables(inductions);
   state_ = LoopUnrollState{loop_induction_variable_, loop->GetLatchBlock(),
-                           loop_condition_block_};
+                           loop_condition_block_, std::move(inductions)};
   for (size_t i = 0; i < factor - 1; ++i) {
     CopyBody(loop, true);
   }
+}
 
-  uint32_t phi_index = GetPhiIndexFromLabel(state_.previous_continue_block_,
-                                            state_.previous_phi_);
-  uint32_t phi_variable =
-      state_.previous_phi_->GetSingleWordInOperand(phi_index - 1);
-  uint32_t phi_label = state_.previous_phi_->GetSingleWordInOperand(phi_index);
-
-  ir::Instruction* original_phi = loop_induction_variable_;
+void LoopUnrollerUtilsImpl::RemoveDeadInstructions() {
+  // Remove the dead phi instructions.
```

Comment:
Assert that these are all Phis being killed?  Also, maybe this could be left to DCE?  Though I'm fine with leaving this here, since it seems to be straightforward.

Reply:
Yeah that was what happened in the original patch but as Alan pointed out the phis aren't just dead but also invalid as the branches will change. Technically it doesn't have to just be phis but that is the main thing being removed, I'll update the comment

Reply:
Ah, yes, please do.  Perhaps renaming the class member to |invalidated_phis_| may make it clearer.